source/Dictionary.cpp

```diff
@@ -92,3 +103,17 @@ double Dictionary::Get(const string &key) const
 {
 	return Get(key.c_str());
 }
+
+
+
+void Dictionary::CheckCollisions() const
+{
+	using element_type = pair<stringAndHash, double>;
+
+	auto equal_found = adjacent_find((*this).begin(), (*this).end()
+			, [](const element_type &a, const element_type &b) {
+		return a.first.GetHash().Get() == b.first.GetHash().Get();
+	});
+	if(equal_found != (*this).end())
+		throw runtime_error { "Found an hash collision on '" + string((*equal_found).first.GetString()) + "'" };
```

Comment:
Although hash collisions using FNV1A are pretty rare, they do happen and it seems less than ideal to crash the game if they happen. The usual solution is to compare the keys, which would amount to a single strcmp per lookup. I wonder if that has a negligible impact of performance or not.

Reply:
For this function to work well, it should be called once a `Dictionary` has been completely filled.
That's the moment where a single call can be made for this kind of checks.
Indeed, after a (shallow) check over the codebase I recognize that Outfit initialization could be the right place to do that call just because it uses `operator[]` and I placed that call just after... But I don't know enough about the codebase (maybe there's a `Dictionary` for ship attribute and one for outfit? I don't know...).

> The usual solution is to compare the keys, which would amount to a single strcmp per lookup. I wonder if that has a negligible impact of performance or not.

I think it has a big impact :P

> it seems less than ideal to crash the game if they happen

:joy: You're right, but the 'reaction' (crashing game through an exception) although exaggerated is good one's: when a collision happens, it is better to not pass any test and report back to programmers as soon as possible, so an action can be taken (changing the seed could suffice).

So the point is: how you would me to report that a collision happened (excluding checks at compile time, they can't be done...) as soon as possible?

Reply:
> Indeed, after a (shallow) check over the codebase I recognize that Outfit initialization could be the right place to do that call just because it uses operator[] and I placed that call just after... But I don't know enough about the codebase (maybe there's a Dictionary for ship attribute and one for outfit? I don't know...).

Yes that is fine. Ship attributes are actually stored internally as an Outfit, and an outfit's attributes are all parsed when Load is done. However, you could still have collisions with an attribute that we check for that isn't present on the outfit. For example, if `"hyperdrive"` and `"abc"` collide and there's an outfit just with `"abc"`, this function will report no problems although at runtime when the game checks for the hyperdrive attribute it will erroneously get `"abc"`.

 >  is better to not pass any test and report back to programmers as soon as possible

That's true, but attributes can be defined by plugin creators as well. So for them this is not exactly a very understandable problem.

Reply:
> Yes that is fine. Ship attributes are actually stored internally as an Outfit, and an outfit's attributes are all parsed when Load is done. However, you could still have collisions with an attribute that we check for that isn't present on the outfit. For example, if "hyperdrive" and "abc" collide and there's an outfit just with "abc", this function will report no problems although at runtime when the game checks for the hyperdrive attribute it will erroneously get "abc".

I will check if there are remnant calls to `Dictionary::operator[]` (which is the only method allowed to insert new keys), but if you say that all attributes are loaded once Load is done, wouldn't it be worth to create a `Dictionary` just to put all of the loaded keys and check?

This will also reduce the number of calls to `Dictionary::CheckCollisions()`...

Where should I begin?

Reply:
> That's true, but attributes can be defined by plugin creators as well. So for them this is not exactly a very understandable problem.

In this case, wouldn't it be good to fail on load (almost a CTD, I know...) but printing or notifying the plugin creator?

Reply:
I think I found the right way, let me know if that's fine for you.